test/test_build_stream.ts

```diff
@@ -0,0 +1,191 @@
+				}
+				const sampleDaemonStreamStep = streamer.next();
+				if (sampleDaemonStreamStep.done) {
+					throw new Error('sample data ended before build stream');
+				}
+				assert.deepEqual(buf.toString(), sampleDaemonStreamStep.value);
+			});
+		});
+
+		// Create a mock tar stream and pipe it to the builder (buildStream)
+		const tarStreamSizeMegaBytes = 1;
+		const tarStreamPromise = mockTarStream(buildStream!, tarStreamSizeMegaBytes);
+
+		return Bluebird.all([
+			tarStreamPromise,
+			buildStreamPromise,
+			mockBuilder.docker.buildImagePromise,
+		]);
+	})
+});
+
+/**
+ * Write the data produced by the given iterator to the given stream,
+ * "spreading" the writes over the node/JS event loop through setImmediate().
+ * This gives stream consumers a chance to consume the data as it is written.
+ * If shouldEndStream is true, stream.end() is called when the iteration ends.
+ * Return a promise that fulfils when the iteration ends.
+ * @param stream The stream to write to
+ * @param iter Iterator that produces data (next().value) to be written
+ * @param shouldEndStream If true, call stream.end() when iteration ends
+ * @returns A promise that fulfils when the iteration ends or rejects on stream errors
+ */
+function eventLoopWriteIterable(
+	stream: Writable,
+	iter: Iterator<string|Buffer>,
+	shouldEndStream: boolean = true,
+): Bluebird<void> {
+	return new Bluebird((resolve, reject) => {
+		stream.on('error', (error) => reject(error));
+
+		function recurseOverEventLoop () {
+			setImmediate(() => {
+				let it = iter.next();
+				if (it.done) {
+					if (shouldEndStream) {
+						stream.end();
+					}
+					resolve();
+				} else {
+					stream.write(it.value);
+					recurseOverEventLoop();
+				}
+			});	
+		}
+		recurseOverEventLoop();
+	});
+}
+
+/**
+ * Create a mock tar stream and write it to the given builder stream in
+ * 1KB writes, with each write spread over the Node/JS event loop.
+ * @param writeStream The builder stream to write to
+ * @returns A promise that fulfils when there is no more data to write
+ */
+function mockTarStream(writeStream: Writable, sizeMegaBytes: number): Bluebird<void> {
+	const mockTarKiloByteBuf = Buffer.allocUnsafe(1024);
+	return eventLoopWriteIterable(writeStream, function*() {
+		for (let i=0; i < sizeMegaBytes * 1024; ++i) {
+			yield mockTarKiloByteBuf;
+		}
+	}());
+}
+
+/**
+ * Dockerode class mock (selected bits)
+ */
+class MockDockerode {
+	
+	public buildImagePromise: Bluebird<void>;
+	public tarStreamMilliseconds: number;
+
+	/**
+	 * Mock of dockerode's buildImage() (not really the same interface though)
+	 * @param inputStream Mock tar stream
+	 * @param outputStream Mock docker build image output stream
+	 * @returns A promise that fulfils when the in/out streams have been fully read/written
+	 */
+	public buildImage(inputStream: Readable): Writable {
+		const outputStream = new Stream.PassThrough();					
+		this.buildImagePromise = new Bluebird((resolve, reject) => {
```

Comment:
```typescript
const outputStream = new Stream.PassThrough();
this.buildImagePromise = this.mockBuildImage(inputStream, outputStream)
return outputStream;
```

Reply:
hehe indeed! I've simplified it further and merged the two methods in one.